### DISCUSSION TITLE
PrimaryEntityId getting set to Guid.Empty when context is being obtained in the Plugin

### DISCUSSION THREAD

**Discussion by vijayrit** [jordimontana82/fake-xrm-easy#565] [Q&A] [answered]
Hi all,

I have a use case where I have to check if the primaryentityid being passed in the context is Guid.Empty or not.

I have written the following test case where I mock the primaryentityid as follows:
`var PluginContext = fakeContext.GetDefaultPluginContext();

            PluginContext.Stage = 10;
            PluginContext.MessageName = "Retrieve";
            PluginContext.BusinessUnitId = Guid.NewGuid();
            PluginContext.CorrelationId = Guid.NewGuid();
            PluginContext.UserId = Guid.NewGuid();
            PluginContext.PrimaryEntityId = Guid.NewGuid();
            PluginContext.PrimaryEntityName = `"isp_userprovision";`

But when I am trying to compare it in the Plugin code I found that it is being set to Guid.Empty

![image](https://user-images.githubusercontent.com/26718999/121654561-7aa10480-cabb-11eb-8b2e-7b89558a6b14.png)

![image](https://user-images.githubusercontent.com/26718999/121654695-9efce100-cabb-11eb-8fac-966716dc26b8.png)

Has someone faced this issue previously? If yes, can you please help me with the solution.

**Comment by BetimBeja** [comment, marked as answer]
Hello @vijayrit 
PrimaryEntityId gets read from the Target entity:
https://github.com/jordimontana82/fake-xrm-easy/blob/f294b9af7319f8e23d4032a9290ec75a628f2435/FakeXrmEasy.Shared/XrmFakedContext.Plugins.cs#L78-L88

**Comment by jordimontana82** [comment]
Yes, this has caused a bit of confusion. When you execute the plugin with the target as a parameter, it'll use the properties of the target as Betim said above, otherwise, you would need to use the method overload that receives the plugin context method instead.

**Reply by vijayrit** [reply]
What if there is no inputparameter in the mocked context? Is this a valid scenario? If yes, how is the primary entity id handled in this case?

**Reply by jordimontana82** [reply]
Could you please post what's the actual ExecutePlugin* method you're using @vijayrit ?

**Reply by jordimontana82** [reply]
If there is no InputParameters then please use this method instead: 

https://github.com/jordimontana82/fake-xrm-easy/blob/f294b9af7319f8e23d4032a9290ec75a628f2435/FakeXrmEasy.Shared/XrmFakedContext.Plugins.cs#L108

**Reply by vijayrit** [reply]
Hi Jordi we are using this method. But we are passing a fakedexecutioncontext for the same. A sample code snippet for reference.
```
public void NegativeScenario()
        {
            var fakeContext = new XrmFakedContext();
            

            var PluginContext = fakeContext.GetDefaultPluginContext();

            var id = Guid.NewGuid();

            // Define Input Parameter
            var inputParameter = new ParameterCollection();


            PluginContext.Stage = 10;
            PluginContext.MessageName = "Retrieve";
            PluginContext.BusinessUnitId = Guid.NewGuid();
            PluginContext.CorrelationId = Guid.NewGuid();
            PluginContext.UserId = Guid.NewGuid();
            PluginContext.PrimaryEntityId = id;
            PluginContext.PrimaryEntityName = "primary entity";

            PluginContext.InputParameters = inputParameter;

            //test data mocked here. Removed for brevity.       
            
            inputParameter.Add("Target", useraccessrecord);

            fakeContext.ExecutePluginWith<UserProvisionDataProvider>(PluginContext);

        }
```

**Reply by BetimBeja** [reply]
You have `inputParameter.Add("Target", useraccessrecord);`. 
Why do you want to override `PluginContext.PrimaryEntityId = id;` and `PluginContext.PrimaryEntityName = "primary entity";` you should use `PluginContext.PrimaryEntityId = useraccessrecord.Id ` and `PluginContext.PrimaryEntityName= useraccessrecord.LogicalName`

**Reply by jordimontana82** [reply]
+1 above. @vijayrit  If you want to pass the Target, then use ExecutePluginWithTarget instead. The ExecutePluginWith<T> method was introduced if you had to access other properties of the context only. Why would you want to pass a Target with a different entity logical name than context.PrimaryEntityName?

**Reply by vijayrit** [reply]
Ok, thanks for the information Jordi and Betim. 

This helps clarify the question. On the same note, is there a documentation of which function to use when, as this would be very helpful in leveraging the full benefits of the Nuget?

**Reply by jordimontana82** [reply]
Thanks, these methods were described here: https://dynamicsvalue.com/get-started/plugins 

As part of the v2 release we'll publish an updated docs page. In the meantime, the page above and the repo itself is the best documentation we have so far...
